app/views.py

```diff
@@ -78,6 +78,106 @@ def admin():
+def admin_services_management():
+    services = services_model.all()
+    return render_template("admin/services/index.html", services=services)
+
+
+@app.route("/admin/services/save", methods=["POST"])
+def admin_services_save():
+    updated_images = None
+    if request.files.getlist("images"):
+        updated_images = util.upload_files(
+            request.files.getlist("images"), app.config["IMAGES_DIR_PATH"]
+        )
+    price = {
+        price[0]: price[1]
+        for price in list(
+            zip(
+                request.form.getlist("price_names[]"),
+                request.form.getlist("price_values[]"),
+            )
+        )
+    }
+    try:
+        service = services_model.get(id=request.form.get("service_id"))
+        # get current set of images
+        images = set(service["images"])
+        # add updated images if any
+        images |= set(updated_images)
+
+        services_model.update(
+            id=service["id"],
+            name=request.form["name"],
+            price=price,
+            description=request.form["description"],
+            images=list(images),
+        )
+        flash("Service has been updated")
+        return redirect(url_for("admin_services_edit", id=service["id"]))
+    except services_model.ErrNotFound:
+        images = updated_images or None
+        services_model.insert(
+            name=request.form["name"],
+            price=price,
+            description=request.form["description"],
+            images=images,
+        )
+        flash("Service has been created")
+        return redirect(url_for("admin_services_add"))
+
+
+@app.route("/admin/services/add")
+def admin_services_add():
+    return render_template("admin/services/upsert.html", mode="add")
+
+
+@app.route("/admin/services/<id>/images/remove", methods=["POST"])
+def admin_services_images_remove(id):
+    try:
+        # remove images from disk if any
+        rm_image = request.form.get("image")
+        util.remove_files([rm_image], app.config["IMAGES_DIR_PATH"])
+
+        service = services_model.get(id=id)
+        # get current set of images and update it
+        service["images"].remove(rm_image)
+
+        services_model.update(
+            id=service["id"],
+            name=service["name"],
+            price=service["price"],
+            description=service["description"],
+            images=service["images"],
+        )
+        flash("Removed image")
+        return redirect(url_for("admin_services_edit", id=service["id"]))
+    except services_model.ErrNotFound:
+        abort(404)
+
+
+@app.route("/admin/services/<id>/edit")
+def admin_services_edit(id: str):
+    try:
+        service = services_model.get(id=id)
+        return render_template(
+            "admin/services/upsert.html", service=service, mode="edit"
+        )
+    except services_model.ErrNotFound:
+        abort(404)
+
+
+@app.route("/admin/services/<id>/delete", methods=["POST"])
+def admin_services_deleting(id):
```

Comment:
small one, just need to update the method name `deleting` -> `delete`

Reply:
I'll update it